File: benchmark/comparison/results/haiku_py/07.py
def safe_head(lst):
    if len(lst) > 0:
        return f"Some({lst[0]})"
    else:
        return "None"

def safe_max(lst):
    if len(lst) > 0:
        return f"Some({max(lst)})"
    else:
        return "None"

def heads_max(lst1, lst2):
    head1 = safe_head(lst1)
    head2 = safe_head(lst2)

    # Extract values
    val1 = None
    val2 = None

    if head1.startswith("Some"):
        val1 = int(head1[5:-1])
    if head2.startswith("Some"):
        val2 = int(head2[5:-1])

    # Return max
    if val1 is not None and val2 is not None:
        return f"Some({max(val1, val2)})"
    elif val1 is not None:
        return f"Some({val1})"
    elif val2 is not None:
        return f"Some({val2})"
    else:
        return "None"

# Test
print(heads_max([5, 1], [3, 2]))
print(heads_max([], [3]))
print(heads_max([3], []))
print(heads_max([], []))

File: benchmark/comparison/results/adv_ts/07.ts
type Option<T> = { tag: "Some"; value: T } | { tag: "None" };

function some<T>(value: T): Option<T> {
  return { tag: "Some", value };
}

function none<T>(): Option<T> {
  return { tag: "None" };
}

function format_option(o: Option<number>): string {
  return o.tag === "Some" ? `Some(${o.value})` : "None";
}

function safe_head(lst: number[]): Option<number> {
  if (lst.length === 0) return none();
  return some(lst[0]);
}

function safe_max(a: Option<number>, b: Option<number>): Option<number> {
  if (a.tag === "Some" && b.tag === "Some") return some(Math.max(a.value, b.value));
  if (a.tag === "Some") return a;
  if (b.tag === "Some") return b;
  return none();
}

function heads_max(a: number[], b: number[]): Option<number> {
  return safe_max(safe_head(a), safe_head(b));
}

console.log(format_option(heads_max([5, 1], [3, 2])));
console.log(format_option(heads_max([], [3])));
console.log(format_option(heads_max([3], [])));
console.log(format_option(heads_max([], [])));
